crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -706,8 +718,44 @@ where
                 is_star,
                 range: _,
             }) => {
+                // Save the state prior to visiting any of the `try` block.
+                //
+                // Potentially none of the `try` block could have been executed prior to executing
+                // the `except` block(s), so we will merge this state with all of the intermediate
+                // states during the `try` block before visiting the `except` blocks.
+                let pre_try_block_state = self.flow_snapshot();
+                let saved_definition_states = std::mem::take(&mut self.try_block_definition_states);
+                let visiting_nested_try_block = self.visiting_try_block;
+
+                // Visit the `try` block!
+                //
+                // If `visiting_try_block == true`, this informs `.add_definition()` to record the state
+                // in `self.try_block_definition_states` after each new definition is recorded.
+                self.visiting_try_block = true;
                 self.visit_body(body);
+                self.visiting_try_block = visiting_nested_try_block;
+
+                // Save the state immediately *after* visiting the `try` block
+                // but *before* we prepare for visiting the `except` block(s).
+                //
+                // We will revert to this state prior to visiting the the `else` block,
+                // as there necessarily must have been 0 `except` blocks executed
+                // if we hit the `else` block.
+                let post_try_block_state = self.flow_snapshot();
+
+                // Prepare for visiting the `except` block(s)
+                let visiting_try_block_states = std::mem::replace(
+                    &mut self.try_block_definition_states,
+                    saved_definition_states,
+                );
```

Comment:
Do we need to do any merging of the `try_block_definition_states` if there are nested try statements?